include/nbl/asset/utils/CPolygonGeometryManipulator.h

```diff
@@ -81,6 +81,81 @@ class NBL_API2 CPolygonGeometryManipulator
 				const_cast<IGeometryBase::SAABBStorage&>(geo->getAABBStorage()) = computeAABB(geo);
 		}
 
+		static inline core::smart_refctd_ptr<ICPUPolygonGeometry> createTriangleListIndexing(ICPUPolygonGeometry* geo, bool makeNewMesh = false)
+		{
+			const auto* indexing = geo->getIndexingCallback();
+			if (!indexing) return nullptr;
+			if (indexing->degree() != 3) return nullptr;
+
+			core::smart_refctd_ptr<ICPUPolygonGeometry> outGeometry;
+			const auto primCount = geo->getPrimitiveCount();
+			const auto maxIndex = geo->getPositionView().getElementCount() - 1;
+			const uint8_t indexSize = maxIndex <= std::numeric_limits<uint16_t>::max() ? sizeof(uint16_t) : sizeof(uint32_t);
+			if (makeNewMesh)
+			{
+				outGeometry = core::move_and_static_cast<ICPUPolygonGeometry>(geo->clone(0u));
+			} else
+			{
+				outGeometry = core::smart_refctd_ptr<ICPUPolygonGeometry>(geo);
+			}
+
+			if (indexing && indexing->knownTopology() == EPT_TRIANGLE_LIST) 
+				return outGeometry;
+
+			auto* outGeo = outGeometry.get();
+			auto indexBuffer = ICPUBuffer::create({ primCount * 3 * indexSize, IBuffer::EUF_INDEX_BUFFER_BIT });
+			auto indexBufferPtr = indexBuffer->getPointer();
+			auto indexView = ICPUPolygonGeometry::SDataView{
+				.composed = {
+					.stride = indexSize,
+				},
+				.src = {
+					.offset = 0,
+					.size = indexBuffer->getSize(),
+					.buffer = std::move(indexBuffer)
+				}
+			};
+
+			if (indexSize == 2)
+			{
+				IPolygonGeometryBase::IIndexingCallback::SContext<uint16_t> context{
+					.indexBuffer = outGeo->getIndexView().getPointer(),
+					.indexSize = indexSize,
+					.beginPrimitive = 0,
+					.endPrimitive = primCount,
+					.out = indexBufferPtr,
+				};
+				indexing->operator()(context);
+
+				indexView.composed.encodedDataRange.u16.minVx[0] = 0;
+				indexView.composed.encodedDataRange.u16.maxVx[0] = maxIndex;
+				indexView.composed.format = EF_R16_UINT;
+				indexView.composed.rangeFormat = IGeometryBase::EAABBFormat::U16;
+
+			} else
+			{
+				IPolygonGeometryBase::IIndexingCallback::SContext<uint32_t> context{
+					.indexBuffer = outGeo->getIndexView().getPointer(),
```

Comment:
the `indexBuffer` in the context is the INPUT, so should be `geo->` not `outGeo->`

It probably works regardless because `outGeo`'s index buffer is same as `geo` until you replace it